### ISSUE TITLE
Multi gpu training slows down every several steps

### ISSUE THREAD

**Issue by xinli94** [horovod/horovod#675]
Hi,

I used horovod to train deepspeech2 model `https://github.com/SeanNaren/deepspeech.pytorch` on 1 server with 4 gpus.

The training slows down significantly every several steps

```
[GPU rank 1] Epoch: [1][30786/34406]    Time 0.783 (2.771)      Data 0.006 (1.204)      Loss 58.0583 (55.2676)
[GPU rank 3] Epoch: [1][30786/34406]    Time 0.782 (2.771)      Data 0.118 (0.686)      Loss 43.3951 (55.5225)
[GPU rank 0] Epoch: [1][30786/34406]    Time 0.782 (2.771)      Data 0.003 (0.416)      Loss 47.8849 (55.1309)
[GPU rank 2] Epoch: [1][30786/34406]    Time 0.782 (2.771)      Data 0.017 (0.021)      Loss 50.1987 (55.5940)
[GPU rank 3] Epoch: [1][30787/34406]    Time 0.636 (2.771)      Data 0.003 (0.686)      Loss 68.1900 (55.5229)
[GPU rank 2] Epoch: [1][30787/34406]    Time 0.636 (2.771)      Data 0.003 (0.021)      Loss 61.7113 (55.5942)
[GPU rank 0] Epoch: [1][30787/34406]    Time 0.636 (2.771)      Data 0.003 (0.416)      Loss 76.5062 (55.1316)
[GPU rank 1] Epoch: [1][30787/34406]    Time 0.637 (2.771)      Data 0.006 (1.204)      Loss 83.6670 (55.2685)
[GPU rank 2] Epoch: [1][30788/34406]    Time 0.628 (2.771)      Data 0.007 (0.021)      Loss 108.1024 (55.5959)
[GPU rank 0] Epoch: [1][30788/34406]    Time 0.628 (2.771)      Data 0.003 (0.416)      Loss 57.8936 (55.1317)
[GPU rank 3] Epoch: [1][30788/34406]    Time 0.630 (2.771)      Data 0.002 (0.686)      Loss 132.4054 (55.5254)
[GPU rank 1] Epoch: [1][30788/34406]    Time 0.641 (2.771)      Data 0.004 (1.204)      Loss 57.8812 (55.2686)
[GPU rank 2] Epoch: [1][30789/34406]    Time 9.185 (2.771)      Data 0.002 (0.021)      Loss 87.6285 (55.5969)
[GPU rank 3] Epoch: [1][30789/34406]    Time 9.184 (2.771)      Data 0.036 (0.686)      Loss 86.9646 (55.5265)
[GPU rank 0] Epoch: [1][30789/34406]    Time 9.185 (2.771)      Data 8.596 (0.416)      Loss 89.1108 (55.1328)
[GPU rank 1] Epoch: [1][30789/34406]    Time 9.173 (2.771)      Data 0.005 (1.204)      Loss 88.9971 (55.2697)
[GPU rank 2] Epoch: [1][30790/34406]    Time 0.869 (2.771)      Data 0.160 (0.021)      Loss 80.5456 (55.5978)
[GPU rank 1] Epoch: [1][30790/34406]    Time 0.867 (2.771)      Data 0.006 (1.204)      Loss 82.8976 (55.2706)
[GPU rank 3] Epoch: [1][30790/34406]    Time 0.869 (2.771)      Data 0.005 (0.686)      Loss 59.3409 (55.5266)
[GPU rank 0] Epoch: [1][30790/34406]    Time 0.869 (2.771)      Data 0.003 (0.416)      Loss 61.3141 (55.1330)
[GPU rank 2] Epoch: [1][30791/34406]    Time 0.718 (2.771)      Data 0.003 (0.021)      Loss 43.5383 (55.5974)
[GPU rank 1] Epoch: [1][30791/34406]    Time 0.718 (2.771)      Data 0.036 (1.204)      Loss 48.2481 (55.2704)
[GPU rank 0] Epoch: [1][30791/34406]    Time 0.718 (2.771)      Data 0.003 (0.416)      Loss 61.0657 (55.1332)
[GPU rank 3] Epoch: [1][30791/34406]    Time 0.736 (2.771)      Data 0.055 (0.686)      Loss 49.9192 (55.5264)
[GPU rank 1] Epoch: [1][30792/34406]    Time 0.662 (2.771)      Data 0.009 (1.204)      Loss 57.3082 (55.2705)
[GPU rank 3] Epoch: [1][30792/34406]    Time 0.644 (2.771)      Data 0.004 (0.686)      Loss 53.6355 (55.5263)
[GPU rank 2] Epoch: [1][30792/34406]    Time 0.662 (2.771)      Data 0.002 (0.021)      Loss 102.2186 (55.5989)
[GPU rank 0] Epoch: [1][30792/34406]    Time 0.663 (2.771)      Data 0.003 (0.416)      Loss 41.6938 (55.1327)
[GPU rank 3] Epoch: [1][30793/34406]    Time 11.944 (2.771)     Data 0.053 (0.686)      Loss 87.1375 (55.5274)
[GPU rank 1] Epoch: [1][30793/34406]    Time 11.945 (2.771)     Data 0.080 (1.204)      Loss 87.3335 (55.2715)
[GPU rank 2] Epoch: [1][30793/34406]    Time 11.945 (2.771)     Data 0.002 (0.021)      Loss 36.2707 (55.5982)
[GPU rank 0] Epoch: [1][30793/34406]    Time 11.945 (2.771)     Data 11.343 (0.416)     Loss 61.0119 (55.1329)
```

But it runs perfectly on one single gpu. Any suggestion on debugging this?

Thanks,
Xin

**Comment by xinli94** [comment]
Also, i checked the log and found that there is no high latency for the first ~10k steps.

**Comment by alsrgv** [comment]
@xinli94, sorry for the delay.  Does deep speech model use sparse gradients, e.g. due to embeddings?   If so, this feature may help - https://github.com/uber/horovod/pull/570.
